charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,302 @@
+            return True
+        return False
+
+    def _apply_k8s_resources(self, force_conflicts: bool = False) -> None:
+        """Apply K8s resources.
+
+        Args:
+            force_conflicts (bool): *(optional)* Will "force" apply requests causing conflicting
+                                    fields to change ownership to the field manager used in this
+                                    charm.
+                                    NOTE: This will only be used if initial regular apply() fails.
+        """
+        self.unit.status = MaintenanceStatus("Creating K8s resources")
+        try:
+            self.k8s_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying K8s resources
+                # re-apply K8s resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying K8s resources")
+                self.logger.warning("Apply K8s resources with forced changes against conflicts")
+                self.k8s_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("K8s resources creation failed") from error
+        try:
+            self.crd_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying CRD resources
+                # re-apply CRD resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying CRD resources")
+                self.logger.warning("Apply CRD resources with forced changes against conflicts")
+                self.crd_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("CRD resources creation failed") from error
+        self.model.unit.status = MaintenanceStatus("K8s resources created")
 
-    def _check_leader(self):
+    def _check_leader(self) -> None:
+        """Check whether a unit is the leader."""
         if not self.unit.is_leader():
             # We can't do anything useful when not the leader, so do nothing.
-            raise CheckFailed("Waiting for leadership", WaitingStatus)
+            self.logger.warning("Not a leader, skipping setup")
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
+
+    def _check_container_connection(self) -> None:
+        """Check if connection can be made with container."""
+        if not self.container.can_connect():
+            raise ErrorWithStatus("Pod startup is not complete", MaintenanceStatus)
+
+    def _check_status(self) -> None:
+        """Check status of workload and set status accordingly."""
+        self._check_leader()
+        container = self.unit.get_container(self._container_name)
+        if container:
```

Comment:
Is there a condition in which `get_container` will return `None`, though? What is this `if` checking?
Also, why not use either the attribute (`self._container`) or property (self.container) that you have defined above?

Reply:
> Atm, container=None will just pass this function

This doesn't really hold, `get_container()` will raise a `ModelError` (which is currently caught in the caller) if the container doesn't exist. So yes, in that sense this `if` doesn't look super important.

> Also, why not use either the attribute (self._container) or property (self.container) that you have defined above?

The only reason I can think of is the case where something has changed since the charm got triggered (and `self._container` was initialised). With that in mind, I guess I should drop `self._container` altogether and use the property to fetch the container info using `self.unit.get_container` every time.

Reply:
> The only reason I can think of is the case where something has changed since the charm got triggered (and self._container was initialised). 

Will the Container object change, though?

> With that in mind, I guess I should drop self._container altogether and use the property to fetch the container info using self.unit.get_container every time.

I think that's appropriate, completely different from the pattern in all other charms, but valid nonetheless.

Reply:
whether to fetch every time or to cache once is awkward - not sure which is better.  If we fetch every time, things can change mid execution.  In either case though, these functions run fast so probably it wont actually happen.

Re `get_container()` never returning `None`, ok sounds good I buy that.  I wonder if a comment in the code is needed here.  Probably not.